File: TicTacToe.Core/Game.cs
// TicTacToe.Core/Game.cs
using System;
using System.Linq;

namespace TicTacToe;

public class Game
{
    private static readonly int[][] _winningLines =
    {
        new[] {0,1,2},
        new[] {3,4,5},
        new[] {6,7,8},
        new[] {0,3,6},
        new[] {1,4,7},
        new[] {2,5,8},
        new[] {0,4,8},
        new[] {2,4,6}
    };

    public Cell[] Board { get; } = new Cell[9];
    public Cell CurrentPlayer { get; private set; } = Cell.X;
    public bool IsGameOver { get; private set; }
    public Cell Winner { get; private set; } = Cell.Empty;

    /// <summary>
    /// Делает ход в ячейку pos (0..8). Возвращает true, если ход принят.
    /// </summary>
    public bool MakeMove(int pos)
    {
        if (IsGameOver || pos < 0 || pos >= Board.Length) return false;
        if (Board[pos] != Cell.Empty) return false;

        Board[pos] = CurrentPlayer;

        if (CheckWinner(CurrentPlayer))
        {
            Winner = CurrentPlayer;
            IsGameOver = true;
        }
        else if (Board.All(c => c != Cell.Empty))
        {
            Winner = Cell.Empty;
            IsGameOver = true;
        }
        else
        {
            CurrentPlayer = CurrentPlayer == Cell.X ? Cell.O : Cell.X;
        }

        return true;
    }

    public void Reset()
    {
        for (int i = 0; i < Board.Length; i++)
            Board[i] = Cell.Empty;

        CurrentPlayer = Cell.X;
        Winner = Cell.Empty;
        IsGameOver = false;
    }

    private bool CheckWinner(Cell player)
    {
        foreach (var line in _winningLines)
        {
            if (Board[line[0]] == player &&
                Board[line[1]] == player &&
                Board[line[2]] == player)
                return true;
        }
        return false;
    }
}

File: TicTacToe.Android/app/src/main/java/com/alexgold25/tictactoe/Game.kt
package com.alexgold25.tictactoe

enum class Cell { E, X, O }

class Game {
    val board: Array<Cell> = Array(9) { Cell.E }
    var current: Cell = Cell.X;     private set
    var winner: Cell = Cell.E;      private set
    var isOver: Boolean = false;    private set

    private val lines = arrayOf(
        intArrayOf(0,1,2), intArrayOf(3,4,5), intArrayOf(6,7,8),
        intArrayOf(0,3,6), intArrayOf(1,4,7), intArrayOf(2,5,8),
        intArrayOf(0,4,8), intArrayOf(2,4,6)
    )

    fun reset() {
        for (i in board.indices) board[i] = Cell.E
        current = Cell.X; winner = Cell.E; isOver = false
    }

    fun makeMove(pos: Int): Boolean {
        if (isOver || pos !in 0..8 || board[pos] != Cell.E) return false
        board[pos] = current
        isOver = checkWinner(current) || board.none { it == Cell.E }
        winner = if (isOver && checkWinner(current)) current else if (isOver) Cell.E else Cell.E
        if (!isOver) current = if (current == Cell.X) Cell.O else Cell.X
        return true
    }

    fun winningLine(): IntArray? {
        if (!isOver || winner == Cell.E) return null
        for (l in lines) {
            val (a,b,c) = l
            if (board[a]==winner && board[b]==winner && board[c]==winner) return l
        }
        return null
    }

    private fun checkWinner(p: Cell): Boolean {
        for (l in lines) {
            val (a,b,c) = l
            if (board[a]==p && board[b]==p && board[c]==p) return true
        }
        return false
    }
}
